Source/Tools/BakeVertexAnimationTexture/Baker.cpp

```diff
@@ -0,0 +1,313 @@
+    outputGeometry->SetIndexBuffer(outputIndexBuffer);
+    outputGeometry->SetDrawRange(TRIANGLE_LIST, 0, indexData.size(), 0, vertexData.size(), false);
+    outputModel->SetGeometry(0, 0, outputGeometry);
+    {
+        auto outputModelFileName = options_.outputFolder_ + GetFileNameAndExtension(options_.inputModel_);
+        auto modelFile = MakeShared<File>(context_, outputModelFileName, FILE_WRITE);
+        if (!outputModel->Save(*modelFile))
+        {
+            Help("Failed to save model file: " + outputModelFileName);
+        }
+    }
+}
+
+void Baker::LoadModel()
+{
+    model_ = MakeShared<Model>(context_);
+    {
+        auto modelFile = MakeShared<File>(context_, options_.inputModel_);
+        if (!model_->Load(*modelFile))
+        {
+            Help("Failed to parse model file: " + options_.inputModel_);
+        }
+    }
+}
+
+void Baker::LoadAnimation()
+{
+    animation_ = MakeShared<Animation>(context_);
+    auto aniFile = MakeShared<File>(context_, options_.inputAnimation_);
+    if (!animation_->Load(*aniFile))
+    {
+        Help("Failed to parse animation file: " + options_.inputAnimation_);
+    }
+    textureHeight_ = NextPowerOfTwo(animation_->GetLength() * options_.targetFramerate_);
+    textureHeight_ = Max(1, Min(textureHeight_, 2048));
+}
+
+inline Vector2 EncodeFloatRG(float v)
+{
+    Vector2 kEncodeMul = Vector2(1.0f, 255.0f);
+    float kEncodeBit = 1.0f / 255.0f;
+    Vector2 enc = kEncodeMul * v;
+
+    enc = enc.Fract();
+    enc.x_ = enc.x_ - enc.y_ * kEncodeBit;
+    return enc;
+}
+inline float DecodeFloatRG(Vector2 enc)
+{
+    Vector2 kDecodeDot = Vector2(1.0f, 1.0f / 255.0f);
+    return enc.DotProduct(kDecodeDot);
+}
+
+void Baker::BuildVAT()
+{
+    auto posLookUp = MakeShared<Image>(context_);
+    posLookUp->SetSize(textureWidth_ * (options_.half_ ? 2 : 1), textureHeight_, 4);
+
+    auto normLookUp = MakeShared<Image>(context_);
+    normLookUp->SetSize(textureWidth_, textureHeight_, 4);
+
+    SharedPtr<Node> node = MakeShared<Node>(context_);
+    auto animatedModel = new AnimatedModel(context_);
+    node->AddComponent(animatedModel, 0);
+    auto animationController = new AnimationController(context_);
+    node->AddComponent(animationController, 0);
+    animatedModel->SetModel(model_);
+    AnimationParameters animationParameters{animation_};
+    animationController->PlayExistingExclusive(animationParameters);
+
+    auto geometry = animatedModel->GetLodGeometry(0, 0);
+    auto vertexBuffer = geometry->GetVertexBuffer(0);
+    VertexBufferReader vbReader{vertexBuffer};
+
+    auto dt = animation_->GetLength() / textureHeight_;
+    auto normalPitch = 4 * normLookUp->GetWidth();
+    auto positionPitch = 4 * posLookUp->GetWidth();
+    for (unsigned y=0; y< textureHeight_; ++y)
+    {
+        auto posData = posLookUp->GetData() + y * positionPitch;
+        auto normData = normLookUp->GetData() + y * normalPitch;
+
+        animatedModel->ApplyAnimation();
```

Comment:
@eugeneko how can I iterate through animation frames in the tool? This code does not update bone node positions.